HIMANI ARORA
22070122079
GARGI MITTAL
22070122064
MEET GOLANI

# Pneumonia Classification using Chest X-rays based on Convolutional Neural Nets with Hyper-parameter tuning for CNN



##Objectives:

1. Study the Pueunmonia dataset.

2. Study CNN in detail with its hyperparameter tuning.

3. Implement the same with google colab.

# Theory:
### Chest X-Rays Dataset
1. Pneumonia is a severe lung infection that causes inflammation in the air sacs of the lungs, leading to serious complications and even death, particularly in children and the elderly. Early detection and diagnosis of pneumonia are critical for effective treatment. One widely used method for diagnosis is through chest X-ray imaging, where radiologists look for signs of infection in the lungs.

2. The dataset provided by Paul Timothy Mooney, titled "Chest X-ray Pneumonia", is a collection of X-ray images categorized into two groups: Pneumonia and Normal.  

3. **Dataset Overview:**

* The dataset, available on Kaggle under the title "Chest X-ray Pneumonia", contains chest X-ray images for the detection of pneumonia.
* These images are classified into two categories:    
Normal: X-ray images without pneumonia.  
Pneumonia: X-ray images with pneumonia.  

* This dataset provides a platform for applying deep learning techniques, particularly Convolutional Neural Networks (CNNs), to develop models that can automatically classify X-ray images and help in the early diagnosis of pneumonia.

4. **Dataset Specifications:**  
Number of Images: Approximately 5,863 X-ray images.

5. **Classes:**    
Normal: Chest X-rays without pneumonia.    
Pneumonia: Chest X-rays with pneumonia, further subdivided into bacterial and viral pneumonia.    
Image Shape: All images are in JPEG format and grayscale.


### **Hyperparameters in Convolutional Neural Networks (CNNs)**

Hyperparameters in Convolutional Neural Networks (CNNs) are configuration settings that define the architecture and training process of the network. Unlike parameters that the network learns during training, such as weights and biases, hyperparameters are set before the training process begins and directly influence the model's performance and computational efficiency.


### **1. Learning Rate (α):**

The learning rate determines how much the model's weights are adjusted in response to the gradient during each backpropagation step. It is one of the most important hyperparameters in any deep learning model.

- **High Learning Rate**: Leads to faster convergence but risks overshooting the optimal solution.
- **Low Learning Rate**: Results in slower convergence, but it allows the model to explore the loss surface more carefully, reducing the risk of overshooting.

The learning rate is typically selected via experimentation, and techniques like **learning rate schedules** or **adaptive learning rates** (used in optimizers like Adam) are commonly used to adjust it dynamically during training.



### **2. Batch Size:**

Batch size defines the number of samples processed before the model's weights are updated.

- **Small Batch Size**: Offers a more granular update of weights and often leads to better generalization. However, it results in noisier gradient estimates and slower training.
- **Large Batch Size**: Allows for faster training due to more stable gradient estimates, but it can lead to suboptimal performance and poor generalization.

Typical values for batch size range from 16 to 128, depending on available computational resources and the size of the dataset.



### **3. Number of Epochs:**

The number of epochs specifies how many complete passes through the training dataset the model will make during training. Too few epochs may result in **underfitting** (where the model doesn't learn enough patterns), while too many epochs may lead to **overfitting** (where the model learns even the noise in the training data).

- **Early Stopping**: This is often used to stop training once the model's performance on the validation set stops improving, thus preventing overfitting.


### **4. Filter Size (Kernel Size):**

In a convolutional layer, the filter or kernel size refers to the height and width of the filters used to perform the convolution operation. Common filter sizes include **3x3**, **5x5**, or **7x7**.

- **Smaller Filters (e.g., 3x3)**: Capture finer details and are typically stacked in deeper layers of the network. They are computationally cheaper and preserve spatial information better.
- **Larger Filters (e.g., 7x7)**: Capture more global features but can lead to loss of fine detail and are computationally more expensive.



### **5. Number of Filters (Depth of Convolutional Layer):**

The number of filters in a convolutional layer (also known as the depth or channels) determines how many feature maps are learned at each layer. Increasing the number of filters allows the model to learn more features, but it also increases computational complexity.

- **Shallow Layers**: Fewer filters (e.g., 32 or 64) are typically used in the initial layers to capture basic features like edges and textures.
- **Deeper Layers**: More filters (e.g., 128, 256, or more) are used in later layers to capture more complex features such as patterns and objects.



### **6. Stride:**

Stride refers to how much the filter moves across the input image during the convolution operation.

- **Stride = 1**: Moves the filter one pixel at a time. This results in highly detailed feature maps but increases the size of the output.
- **Stride > 1**: Reduces the spatial dimensions of the feature maps by skipping pixels. This reduces the computational load but may result in loss of spatial information.

Common stride values are 1 or 2, depending on the desired output size and computational constraints.



### **7. Padding:**

Padding is the process of adding extra pixels (usually zeros) around the edges of the input image before applying the convolution operation. Padding is used to control the output size of the convolutional layer.

- **Valid Padding (No Padding)**: Reduces the spatial dimensions of the input image by not adding any padding. This often leads to smaller feature maps.
- **Same Padding**: Adds padding in such a way that the output size is the same as the input size, which allows for more control over the depth of the network.



### **8. Pooling Size:**

Pooling layers (e.g., max pooling or average pooling) are used to downsample the feature maps, reducing the spatial dimensions while retaining important features.

- **Max Pooling**: Takes the maximum value from a patch of the feature map.
- **Average Pooling**: Takes the average of the values in the patch.

Pooling sizes are usually 2x2 or 3x3, and pooling helps in reducing the number of parameters and computation while also controlling overfitting.



### **9. Dropout Rate:**

Dropout is a regularization technique that helps prevent overfitting by randomly setting a fraction of the neurons to zero during each training iteration.

- **Dropout Rate (p)**: Specifies the fraction of neurons to drop (e.g., p = 0.5 means 50% of the neurons are randomly ignored during each training step).
- **Common Values**: Dropout rates typically range from 0.2 to 0.5, with higher dropout rates used in dense layers to improve generalization.


### **10. Optimizer:**

The optimizer determines how the network's weights are updated during training. Common optimizers include:

- **SGD (Stochastic Gradient Descent)**: The most basic optimizer, with momentum to help accelerate the convergence.
- **Adam (Adaptive Moment Estimation)**: Combines the benefits of SGD with adaptive learning rates and momentum, making it a popular choice for most deep learning models.
- **RMSProp**: Another adaptive learning rate optimizer that works well for recurrent neural networks (RNNs) but is also useful for CNNs.

The choice of optimizer can significantly affect training speed and convergence.


### **11. Activation Function:**

The activation function introduces non-linearity into the network, allowing it to learn complex patterns.

- **ReLU (Rectified Linear Unit)**: The most commonly used activation function in CNNs, defined as \( f(x) = \max(0, x) \). It speeds up convergence and mitigates the vanishing gradient problem.
- **Sigmoid**: Often used in the output layer for binary classification, but less common in hidden layers due to slower convergence.
- **Softmax**: Typically used in the output layer for multi-class classification problems to normalize output values into probabilities.



### **12. Weight Initialization:**

Weight initialization helps to prevent issues like vanishing or exploding gradients. Common initialization methods include:

- **He Initialization**: Ideal for ReLU activation functions.
- **Xavier Initialization**: Useful for tanh or sigmoid activations.

Proper weight initialization can lead to faster convergence during training.



Hyperparameters in CNNs play a crucial role in determining the model's performance and efficiency. These parameters, including learning rate, batch size, number of epochs, filter size, and more, need to be carefully tuned to strike the right balance between learning capacity and computational efficiency. By conducting hyperparameter optimization, models can achieve better accuracy, reduced loss, and improved generalization to unseen data.

In [ ]:
# Install Kaggle API
!pip install kaggle -qq
# Make a directory for Kaggle and upload the kaggle.json file
!mkdir -p ~/.kaggle
!cp kaggle.json ~/.kaggle/
# Set permissions for the Kaggle API token
!chmod 600 ~/.kaggle/kaggle.json
# Download the Pneumonia dataset from Kaggle
!kaggle datasets download -d paultimothymooney/chest-xray-pneumonia --unzip


Dataset URL: https://www.kaggle.com/datasets/paultimothymooney/chest-xray-pneumonia
License(s): other
100% 2.29G/2.29G [01:49<00:00, 17.2MB/s]
100% 2.29G/2.29G [01:49<00:00, 22.4MB/s]


In [ ]:
!pip install keras-tuner -qq

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 129.1/129.1 kB 10.0 MB/s eta 0:00:00


In [ ]:
import tensorflow as tf
from tensorflow.keras.preprocessing.image import ImageDataGenerator
train_dir = '/content/chest_xray/train'
val_dir = '/content/chest_xray/val'
test_dir = '/content/chest_xray/test'

In [ ]:
# Image data generators for data augmentation
train_datagen = ImageDataGenerator(
rescale=1./255,
rotation_range=40,
width_shift_range=0.2,
height_shift_range=0.2,
shear_range=0.2,
zoom_range=0.2,
horizontal_flip=True,
fill_mode='nearest'
)
val_datagen = ImageDataGenerator(rescale=1./255)
test_datagen = ImageDataGenerator(rescale=1./255)
# Data generators
train_generator = train_datagen.flow_from_directory(train_dir,
target_size=(150, 150),
batch_size=32,
class_mode='binary'
)
val_generator = val_datagen.flow_from_directory(
val_dir,
target_size=(150, 150),
batch_size=32,
class_mode='binary'
)
test_generator = test_datagen.flow_from_directory(
test_dir,
target_size=(150, 150),
batch_size=32,
class_mode='binary'
)

Found 5216 images belonging to 2 classes.
Found 16 images belonging to 2 classes.
Found 624 images belonging to 2 classes.


In [ ]:
import keras_tuner as kt
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Dropout

def build_model(hp):
  model = Sequential()
  # Convolutional layers
  model.add(Conv2D(
  filters=hp.Int('conv_1_filters', min_value=32, max_value=128, step=32),
  kernel_size=(3, 3),
  activation='relu',
  input_shape=(150, 150, 3)
  ))
  model.add(MaxPooling2D((2, 2)))
  model.add(Conv2D(
  filters=hp.Int('conv_2_filters', min_value=64, max_value=256, step=64),
  kernel_size=(3, 3),
  activation='relu'
  ))
  model.add(MaxPooling2D((2, 2)))
  model.add(Conv2D(
  filters=hp.Int('conv_3_filters', min_value=128, max_value=512, step=128),
  kernel_size=(3, 3),
  activation='relu'
  ))
  model.add(MaxPooling2D((2, 2)))
  model.add(Flatten())
  model.add(Dense(
  units=hp.Int('dense_units', min_value=128, max_value=512, step=64),
  activation='relu'
  ))
  model.add(Dropout(hp.Float('dropout_rate', min_value=0.2, max_value=0.5, step=0.1)))
  model.add(Dense(1, activation='sigmoid'))
  model.compile(
  optimizer=hp.Choice('optimizer', values=['adam', 'sgd']),
  #optimizer=tf.keras.optimizers.Adam(learning_rate=learning_rate),
  loss='binary_crossentropy',
  metrics=['accuracy']
  )
  return model


This code defines a CNN model for hyperparameter tuning using Keras Tuner. It has three convolutional layers, with the number of filters for each layer being tunable between defined ranges (`conv_1_filters`, `conv_2_filters`, `conv_3_filters`). After each convolutional layer, there's a max pooling layer. The model is flattened and followed by a dense layer, where the number of units (`dense_units`) and dropout rate (`dropout_rate`) are tunable. The optimizer is also tunable, allowing for either 'adam' or 'sgd', and the model is compiled for binary classification with the `binary_crossentropy` loss function.

In [ ]:
tuner = kt.RandomSearch(
  build_model,
  objective='val_accuracy',
  max_trials=10, # Number of different hyperparameter combinations to try
  executions_per_trial=3, # Number of models to build and evaluate per trial
  directory='my_dir',
  project_name='pneumonia_cnn',
  seed = 2
)

stop_early = tf.keras.callbacks.EarlyStopping(monitor='val_loss', patience=5)

# Search for the best hyperparameters
tuner.search(
  train_generator,
  epochs=3,
  validation_data=val_generator,
  callbacks=[stop_early]
)

# Get the best model and hyperparameters
best_model = tuner.get_best_models(num_models=1)[0]
best_hyperparameters = tuner.get_best_hyperparameters(num_trials=1)[0]

print(f"Best Hyperparameters: {best_hyperparameters.values}")

Trial 10 Complete [00h 12m 11s]
val_accuracy: 0.8541666666666666

Best val_accuracy So Far: 0.875
Total elapsed time: 02h 05m 06s
Best Hyperparameters: {'conv_1_filters': 128, 'conv_2_filters': 64, 'conv_3_filters': 128, 'dense_units': 384, 'dropout_rate': 0.4, 'optimizer': 'adam'}


/usr/local/lib/python3.10/dist-packages/keras/src/saving/saving_lib.py:576: UserWarning: Skipping variable loading for optimizer 'adam', because it has 2 variables whereas the saved optimizer has 22 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


In [ ]:
test_loss, test_acc = best_model.evaluate(test_generator)
print(f'Test Accuracy: {test_acc}')
best_model.save('best_pneumonia_cnn_model.h5')

20/20 ━━━━━━━━━━━━━━━━━━━━ 11s 441ms/step - accuracy: 0.6666 - loss: 0.6589


Test Accuracy: 0.6538461446762085


The code executes multiple runs with different hyperparameter combinations, aiming to optimize the model's performance. After running the models for 3 epochs over 4 iterations, it saves the best-performing model based on validation accuracy. While the validation accuracy is about 85%, the testing accuracy is significantly lower at 65%, possibly indicating overfitting or a mismatch in data representation between the training and testing datasets.

### Inferences:
1. **Overfitting Likely Occurred**: The large gap between validation accuracy (85%) and testing accuracy (65%) suggests that the model may have learned patterns specific to the training/validation data that do not generalize well to unseen test data.
   
2. **Potential Data Issues**: The discrepancy between validation and testing accuracy could also imply issues with the data split or sampling methods, where the test set may not accurately represent the distribution of the training/validation set.

3. **Model Performance on Test Data**: Despite identifying hyperparameters that perform well on validation data, the lower testing accuracy reveals that the model may still need improvements, such as regularization, better data preprocessing, or different architectures.

4. **Ideal Hyperparameters Identified for the Validation Set**: Regardless of the test set performance, the experimentation has successfully identified a set of hyperparameters that consistently return the best validation accuracy for this particular model and dataset.

5. **Future Improvements**: Addressing overfitting through techniques like cross-validation, dropout, or data augmentation may help improve generalization to the test data.